### PR TITLE
[EIC-1103]: fixed pdf search by adding custom search algo

### PR DESCRIPTION
**Pull request by danielkuti09**
closes https://github.com/CRJI/EIC/issues/1103
closes https://github.com/CRJI/EIC/issues/1104

- created custom search algorithm for pdf search only and extracted in a separate utility function in the same layers folder
- added unit test for the custom algorithm
- temporarily changed pdf searching to synchronous searching due to a bug that affected the highlight navigation.

### CODE REVIEW

File: src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { generateHighlights, flattenDOMNodes } from '../index'
+
+describe('PDF viewer utilities', () => {
+    it('checks if all marks match the query', () => {
+        for (let index = 1; index < 3; index++) {
+            document.body.innerHTML = '' // reset document.body between files / use cases
+            const fileContent = require('fs').readFileSync(path.join(__dirname, `./${index}.html`))
+            document.body.innerHTML = fileContent
+
+            const marks = generateHighlights([{ pageNum: 0, index: 0 }], document.body.children[0], 'metod', { pageNumber: 0 }, 0)
```
Comment (gabriel-v) [inline src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js:16]: Can we simplify this and remove all the `[{ pageNum: 0, index: 0 }]` and `{ pageNumber: 0 }, 0` from this function interface? A lot of zeros there...

I think it might be a good idea to separate adding the marks to a page (implemented here) from setting and changing the selected orange mark that you scroll with next/prev.

That way, this test is cleaner, and the next/prev button doesn't need to re-run the matcher algorithm for each button press, just find and switch the orange mark to next/prev position. Which can be tested separately with some simple data.

File: src/components/pdf-viewer/layers/util/generateHighlights/index.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    let highlighIndex = 0
+    const currentPageSearchResults = searchResults.filter((match) => match.pageNum === page.pageNumber)
+    if (!currentPageSearchResults.length) return []
+
+    const nodes = flattenDOMNodes(container.children)
+    const { matchedIndexes, tuples } = generateHighlightsData(nodes, query)
+    if (!matchedIndexes.length) return []
+
+    const createMark = (text: string): string =>
+        `<mark ${currentPageSearchResults[highlighIndex]?.index === currentHighlightIndex ? 'class="active"' : ''} id="highlight-${
```
Comment (gabriel-v) [inline src/components/pdf-viewer/layers/util/generateHighlights/index.ts:65]: see above; could we separate setting the orange highlight from creating and moving the mark around?

**Review by gabriel-v — CHANGES_REQUESTED** [review]
First and second tests have significant copy/paste so let's combine them into one big test that does everything.

I would test all words on all pages, so we get a good number of empty results too.

So it would go like this:
- for each page, read HTML, extract text, extract words, save all in vars
- combine and de-duplicate words
- for each (word, page) combination, verify the marked results

**Comment by gabriel-v** [comment]
Btw CI is failing - pls commit the new package that was installed for testing


https://jenkins.liquiddemo.org/liquidinvestigations/hoover-ui/2692/2/4

```
Test environment jest-environment-jsdom cannot be found. 

As of Jest 28 "jest-environment-jsdom" is no longer shipped by default, make sure to install it separately.
```

File: src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    it('checks if all marks match the query for multiple words', () => {
+        document.body.innerHTML = '' // reset document.body between files / use cases
+        const string = require('fs').readFileSync(path.join(__dirname, `./3.html`))
+        document.body.innerHTML = string
+
+        const nodes = flattenDOMNodes(document.body.children)
+        const words = nodes
+            .map((node) => node.innerHTML?.trim().split(/\s+/))
+            .flat()
+            .filter(Boolean)
```
[the remark this answers is not in the document]
Comment (danielkuti09) [reply src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js:33]: 'metoda'
'operaţională'
'(transformarea'
'laplace)'
'38.1.'

We already have words like the ones above, and many other edge cases extracted and used. Special characters, diacritics, numbers, dots, and so on. All part of the 500+ words we are running the test agains.

File: src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js (hunk 1 above)
Comment (danielkuti09) [reply src/components/pdf-viewer/layers/util/generateHighlights/__tests__/generateHighlights.test.js:16]: Very good idea, thanks! Done, and highlight navigation is faster now.

**Review by gabriel-v — APPROVED** [review]
